src/hotspot/share/cds/heapShared.hpp

```diff
@@ -138,21 +139,90 @@ class ArchivedKlassSubGraphInfoRecord {
 
+struct LoadedArchiveHeapRegion;
+
 class HeapShared: AllStatic {
   friend class VerifySharedOopClosure;
- private:
 
+public:
+  // At runtime, heap regions in the CDS archive can be used in two different ways,
+  // depending on the GC type:
+  // - Mapped: (G1 only) the regions are directly mapped into the Java heap
+  // - Loaded: At VM start-up, the objects in the heap regions are copied into the
+  //           Java heap. This is easier to implement than mapping but
+  //           slightly less efficient, as the embedded pointers need to be relocated.
+  static bool can_use() { return can_map() || can_load(); }
+
+  // Can this VM write heap regions into the CDS archive? Currently only G1+compressed{oops,cp}
+  static bool can_write() {
+    assert(DumpSharedSpaces, "only when writing static archive");
+    CDS_JAVA_HEAP_ONLY(return (UseG1GC && UseCompressedOops && UseCompressedClassPointers);)
+    NOT_CDS_JAVA_HEAP(return false;)
+  }
+
+  // Can this VM map archived heap regions? Currently only G1+compressed{oops,cp}
+  static bool can_map() {
+    CDS_JAVA_HEAP_ONLY(return (UseG1GC && UseCompressedOops && UseCompressedClassPointers);)
+    NOT_CDS_JAVA_HEAP(return false;)
+  }
+  static bool is_mapped() {
+    return closed_regions_mapped() && open_regions_mapped();
+  }
+
+  // Can this VM load the objects from archived heap regions into the heap at start-up?
+  static bool can_load()  NOT_CDS_JAVA_HEAP_RETURN_(false);
+  static void verify_loaded_heap() NOT_CDS_JAVA_HEAP_RETURN;
+  static bool is_loaded() {
+    CDS_JAVA_HEAP_ONLY(return _is_loaded;)
+    NOT_CDS_JAVA_HEAP(return false;)
+  }
+
+  static bool are_archived_strings_available() {
+    return is_loaded() || closed_regions_mapped();
+  }
+  static bool are_archived_mirrors_available() {
+    return is_fully_available();
+  }
+  static bool is_fully_available() {
+    return is_loaded() || is_mapped();
+  }
+
+private:
 #if INCLUDE_CDS_JAVA_HEAP
   static bool _closed_regions_mapped;
   static bool _open_regions_mapped;
+  static bool _is_loaded;
   static DumpedInternedStrings *_dumped_interned_strings;
 
+  // Support for loaded archived heap. These are cached values from
+  // LoadedArchiveHeapRegion's.
+  static uintptr_t _dumptime_base_0;
+  static uintptr_t _dumptime_base_1;
+  static uintptr_t _dumptime_base_2;
+  static uintptr_t _dumptime_base_3;
+  static uintptr_t _dumptime_top;
+  static intx _runtime_offset_0;
+  static intx _runtime_offset_1;
+  static intx _runtime_offset_2;
+  static intx _runtime_offset_3;
+  static uintptr_t _loaded_heap_bottom;
+  static uintptr_t _loaded_heap_top;
+  static bool _loading_failed;
+
 public:
   static unsigned oop_hash(oop const& p);
   static unsigned string_oop_hash(oop const& string) {
     return java_lang_String::hash_code(string);
   }
 
+  static bool load_heap_regions(FileMapInfo* mapinfo);
+  static bool is_in_loaded_heap(uintptr_t o) {
+    return (_loaded_heap_bottom <= o && o < _loaded_heap_top);
+  }
```

Comment:
This could be a private function since it is only referenced at line 223.

Reply:
Fixed.